### ISSUE TITLE
Automatic offline machine translation

### ISSUE THREAD

**Issue by badonyx** [damus-io/damus#1478]
## user story

As a Damus user who would like to read notes in foreign languages, I would like my local device to translate notes for me automatically, so that I can seamlessly enjoy reaading notes in foreign languages on my device translated to my preferred language.

### acceptance criteria
1. there is a method to enable offline local device auto translation (i.e. no need to tap "translate note")
2. there is a way to disable
3. User is warned if there is a need to download offline language packs 
4. data size download MB/GB is displayed
5. If not on wifi warn user about mobile network bandwidth

Machine translation of notes could be done on-device or otherwise using the language API provided by Apple (which can do offline translations if you download additional data in system settings). ~~Unfortunately Apple does not seem to allow access to their translation API?~~ (Edit: I have seen some activitypub app do this with a button that calls the system translation UI on the note contents, so it seems that is possible at least.)

A nice alternative I found is Google's [ML Kit](https://developers.google.com/ml-kit/language/translation). It is used in another nostr client [nostrmo](https://github.com/haorendashu/nostrmo). The translation quality is good and the downloaded models are not very large -- on nostrmo I have loaded 6 languages for translation to English and the app's total documents & data size is still only ~300 MB.

**Comment by tyiu** [comment]
I would love to have offline translations and have spent a significant amount of time investigating. However,

1. Apple has not opened up their translation APIs.
2. Google's ML Kit for iOS is poorly maintained and does not work for modern development environments. They require Pods but the industry standard is to use Swift Package Manager. It also doesn't work natively for the new Apple M* processors.

I hope things will develop over the next year or so, but nothing is ready at the moment.

**Comment by jb55** [comment]
Let's keep this open, maybe a solution will crop up eventually

**Comment by badstacker** [comment]
> some activitypub app do this with a button that calls the system translation UI on the note contents

I have no idea how this was implemented, but it would be a great alternative.

**Comment by badstacker** [comment]
> I will try to find the app again if it helps.

It is [Radiant](https://radiant.social/).

**Comment by alltheseas** [comment]
@badstacker is radiant FOSS?

**Comment by badstacker** [comment]
I do not think it is FOSS.

**Comment by alltheseas** [comment]
@alexgleason any idea how Radiant mastodon client does iOS translations?

**Comment by tyiu** [comment]
Whoa, that's definitely Apple native translations. I wonder if they're somehow hacking together some UIKit functionality to get that to work. I'll dig around a bit.

**Comment by alltheseas** [comment]
@tyiu just found this on iOS. Have you explored? 

![image](https://github.com/damus-io/damus/assets/64376233/48c64260-9deb-432b-8c48-5b80dbf948f2)

**Comment by tyiu** [comment]
> @tyiu just found this on iOS. Have you explored?

I've seen that but I don't think that has anything to do with what Radiant does to expose the Translate sheet.

@bryanmontz do you know how Radiant might be accessing Apple's native translate sheet in-app? Is there some UIKit or accessibility magic going on? I don't think there's an API for it at the moment.

**Comment by bryanmontz** [comment]
No, I'm not sure how they're triggering Apple's translate sheet, if that's what's happening. I googled around for a minute and didn't see a way to do it.

**Comment by alltheseas** [comment]
Support@radiant.social

Jake is the dev name. Maybe you could ask him @tyiu

**Comment by badstacker** [comment]
> I wonder if they're somehow hacking together some UIKit functionality to get that to work.

@tyiu I think you are on the right track here. Notice in my screenshot those two big dots at the beginning of the text -- those only appear after translation, and they sort of look like a collapsed form of the text selection cursors. Suggests they are doing some trickery to select the text and then trigger the native translation somehow.

**Comment by tyiu** [comment]
> [Support@radiant.social](mailto:Support@radiant.social)
> 
> Jake is the dev name. Maybe you could ask him @tyiu

Seems like @jknlsn might be the developer. I've privately messaged him on Mastodon to see if he'd be willing to offer insights.

**Comment by alltheseas** [comment]
> > I wonder if they're somehow hacking together some UIKit functionality to get that to work.
> 
> @tyiu I think you are on the right track here. Notice in my screenshot those two big dots at the beginning of the text -- those only appear after translation, and they sort of look like a collapsed form of the text selection cursors. Suggests they are doing some trickery to select the text and then trigger the native translation somehow.

Interesting approach!

Here is the manual flow
![Screenshot 2024-01-18 at 6 36 23 PM](https://github.com/damus-io/damus/assets/64376233/8deefd0d-e00d-42f1-9322-38a519601bcb)

Screenshots:

**Comment by alltheseas** [comment]
Here is the suggested auto-translate workflow, given you can simulate the manual actions @tyiu 

![Screenshot 2024-01-18 at 6 38 13 PM](https://github.com/damus-io/damus/assets/64376233/c0b1832c-5cd7-4f1c-bb90-cb74bdaf037f)

**Comment by alltheseas** [comment]
Changing name to automatic offline machine translation

@tyiu solved the underlying technical challenge with new iOS translate API

https://github.com/damus-io/damus/issues/2302

**Comment by alltheseas** [comment]
added user story to OP's top entry on this issue. looks like iOS 18 is required for this.

**Comment by alltheseas** [comment]
pending #2332 fix

**Comment by tkhumush** [comment]
This will be a very interesting feature. 
iOS18.1 should enable this feature.

**Comment by alltheseas** [comment]
@tyiu any idea if iOS 18 introduced iOS translation regression?

![image](https://github.com/user-attachments/assets/72ada30c-b619-4e5c-8078-f3eaf20f3dd2)

**Comment by jb55** [comment]
we have this now right?
